### PR TITLE
Removes AI and Charge parts from tablets, adds support for more later.

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
## About The Pull Request

Mostly a copy paste from commit comments:

### Charge parts

Removes charge parts from tablets entirely
Machine computers now directly use power from the machine they're in, while the rest uses power cells.
Silicon tablets don't use power at all, since they are needed to see and state laws, we don't want it running out of power.

### AI stuff

Removes the intellicard slot, moves it to the AI restorer app
AI restorer app now requires RD access instead of Bridge access. No access is needed to open it, so RD's console is still an option. This is because the app will make it possible to restore AI's from anywhere now that it doesn't require unique parts, so tying it to at least RD access means you need to use a console that has the app pre-installed, or get the RD/Captain to install it, keeping those limitations sorta.
Adds support for apps holding stuff (currently only used in AI restorer holding intellicards)
Calls kill_program on apps that are being deleted, so we can clear references or drop held things on an app's deletion

### Examine stuff

Moves examine to the modular tablet
Machine computers now examine the processor inside of it
Fixes examine text for tablet apps (only Nuke op app, now additionally AI restorer)

### Misc tablet stuff

Removes 'advanced' tablet subtypes that we used before PDAs were added, in some jobs.
Replaces Roboticist's advanced tablet mail with a laptop
Moves the notepad's note var from the tablet, to the note app
Moves modular computer's defines into their own file
Removes ``is_modular_computer()`` and ``get_modular_computer_parts_examine()``
Replaces PDA's ``insert_type`` with just checking if they should have a pen roundstart or not.

## Why It's Good For The Game

Currently my focus is to remove all PC parts, when that's done we can remove all the code that it requires, making tablets run much better and be way less confusing than it currently is.
I should write a hackmd to explain my end goal here, but in a single sentence: 'simplify PDAs so it's not as confusing to work with'. I'll write it when I finish removing all PC parts, which I did mention some should be removed in my old hackmd https://hackmd.io/-523YJ5qRqigIS_zqns2pA but I think it's better if we gut it all entirely. Once that's done things will be more clear to write what I would like to do with it.

## Changelog

:cl:
balance: AI restorer app now requires RD access to download.
fix: Nukie tablets now properly show when a nuke is going off in their examine.
balance: Tablet's charge and AI holder parts have been removed entirely. Tablets now only have a cell, and computers only use their area's APC.
code: Modular computer's power handling, the AI restorer app's functionality, and modular computer's examining has been refactored, please report any bugs you might find.
/:cl:

### CODE REVIEW

File: code/modules/modular_computers/file_system/programs/airestorer.dm [hunk 1]
```diff
@@ -1,127 +1,139 @@
...
+	data["AI_present"] = !!stored_card?.AI
 	data["error"] = null
-	if(!aicard)
+
+	if(!stored_card)
 		data["error"] = "Please insert an intelliCard."
+	else if(!stored_card.AI)
+		data["error"] = "No AI located"
+	else if(stored_card.flush)
+		data["error"] = "Flush in progress"
```
Comment (san7890) [inline code/modules/modular_computers/file_system/programs/airestorer.dm:131]:
```suggestion
	else if(!stored_card.AI)
		data["error"] = "No AI located!"
	else if(stored_card.flush)
		data["error"] = "Flush in progress..."
```

Comment (JohnFulpWillard) [reply code/modules/modular_computers/file_system/programs/airestorer.dm:131]: this is the opposite bro..
no ai isn't something to exclaim about, it's just a generic error
AI being deleted as we speak however, is something to exclaim.

**Comment by msgerbs** [comment]
Maybe I missed something but why are we removing computer parts?

File: code/modules/modular_computers/file_system/programs/airestorer.dm (hunk 1 above)
Comment (san7890) [reply code/modules/modular_computers/file_system/programs/airestorer.dm:131]: i just thought some punctuation was better than nothing

**Comment by JohnFulpWillard** [comment]
> Maybe I missed something but why are we removing computer parts?

They make tablets too confusing for players and it isn't fun to code things around them, I talked about it a little more in https://github.com/tgstation/tgstation/pull/66924

**Comment by private-tristan** [comment]
> They make tablets too confusing for players and it isn't fun to code things around them, I talked about it a little more in 66924

is there even a reason why they should be called modular computers now that all the 'modular' stuff is being removed?

**Comment by JohnFulpWillard** [comment]
> > They make tablets too confusing for players and it isn't fun to code things around them, I talked about it a little more in 66924
> 
> is there even a reason why they should be called modular computers now that all the 'modular' stuff is being removed?

i guess the apps lol

**Comment by Mothblocks** [comment]
Correct

File: code/modules/modular_computers/computers/item/computer.dm [hunk 2]
```diff
@@ -295,7 +295,33 @@ GLOBAL_LIST_EMPTY(TabletMessengers) // a list of all active messengers, similar
...
+	var/obj/item/computer_hardware/hard_drive/hdd = all_components[MC_HDD]
+	for(var/datum/computer_file/app_examine as anything in hdd.stored_files)
+		if(app_examine.on_examine(src, user))
+			. += app_examine.on_examine(src, user)
+
+	var/obj/item/computer_hardware/card_slot/card_slot = all_components[MC_CARD]
+	var/obj/item/computer_hardware/card_slot/card_slot2 = all_components[MC_CARD2]
+	var/multiple_slots = istype(card_slot) && istype(card_slot2)
+	if(card_slot)
+		if(card_slot?.stored_card || card_slot2?.stored_card)
```
Comment (Mothblocks) [inline code/modules/modular_computers/computers/item/computer.dm:307]:
```suggestion
		if(card_slot.stored_card || card_slot2?.stored_card)
```

File: code/modules/modular_computers/computers/item/computer.dm [hunk 3]
```diff
@@ -295,7 +295,33 @@ GLOBAL_LIST_EMPTY(TabletMessengers) // a list of all active messengers, similar
...
+			var/obj/item/card/id/first_ID = card_slot?.stored_card
+			var/obj/item/card/id/second_ID = card_slot2?.stored_card
```
Comment (Mothblocks) [inline code/modules/modular_computers/computers/item/computer.dm:309]: Just `first_id` and `second_id`.

File: code/modules/modular_computers/computers/item/computer.dm [hunk 4]
```diff
@@ -295,7 +295,33 @@ GLOBAL_LIST_EMPTY(TabletMessengers) // a list of all active messengers, similar
...
+			var/multiple_cards = istype(first_ID) && istype(second_ID)
+			if(Adjacent(user))
+				. += "It has [multiple_slots ? "two slots" : "a slot"] for identification cards installed[multiple_cards ? " which contain [first_ID] and [second_ID]" : ", one of which contains [first_ID ? first_ID : second_ID]"]."
```
Comment (Mothblocks) [inline code/modules/modular_computers/computers/item/computer.dm:312]:
```suggestion
				. += "It has [multiple_slots ? "two slots" : "a slot"] for identification cards installed[multiple_cards ? " which contain [first_ID] and [second_ID]" : ", one of which contains [first_ID || second_ID]"]."
```

File: code/modules/modular_computers/computers/item/computer.dm [hunk 5]
```diff
@@ -295,7 +295,33 @@ GLOBAL_LIST_EMPTY(TabletMessengers) // a list of all active messengers, similar
...
+				. += "It has [multiple_slots ? "two slots" : "a slot"] for identification cards installed, [multiple_cards ? "both of which appear" : "and one of them appears"] to be occupied."
+			. += span_info("Alt-click [src] to eject the identification card[multiple_cards ? "s":""].")
+		else
+			. += "It has [multiple_slots ? "two slots" : "a slot"] installed for identification cards."
+
+	var/obj/item/computer_hardware/printer/printer_slot = all_components[MC_PRINT]
+	if(printer_slot)
+		. += "It has a printer installed."
+		if(Adjacent(user))
+			. += "The printer's paper levels are at: [printer_slot.stored_paper]/[printer_slot.max_paper].</span>"
```
Comment (Mothblocks) [inline code/modules/modular_computers/computers/item/computer.dm:323]: I forget, is it normal for none of these to have span_notice? Yes, right?

File: code/modules/modular_computers/computers/item/computer.dm [hunk 6]
```diff
@@ -426,15 +446,11 @@ GLOBAL_LIST_EMPTY(TabletMessengers) // a list of all active messengers, similar
...
 /obj/item/modular_computer/proc/alert_call(datum/computer_file/program/caller, alerttext, sound = 'sound/machines/twobeep_high.ogg')
 	if(!caller || !caller.alert_able || caller.alert_silenced || !alerttext) //Yeah, we're checking alert_able. No, you don't get to make alerts that the user can't silence.
-		return
+		return FALSE
 	playsound(src, sound, 50, TRUE)
-	visible_message(span_notice("The [src] displays a [caller.filedesc] notification: [alerttext]"))
-	var/mob/living/holder = loc
-	if(istype(holder))
-		to_chat(holder, "[icon2html(src)] [span_notice("The [src] displays a [caller.filedesc] notification: [alerttext]")]")
+	visible_message(span_notice("[icon2html(src)] [span_notice("The [src] displays a [caller.filedesc] notification: [alerttext]")]"))
```
Comment (Mothblocks) [inline code/modules/modular_computers/computers/item/computer.dm:453]:
```suggestion
	visible_message(span_notice("[icon2html(src)] [span_notice("[src] displays a [caller.filedesc] notification: [alerttext]")]"))
```

I think this is a bug?

File: code/modules/modular_computers/computers/item/computer_power.dm [hunk 7]
```diff
@@ -59,6 +54,10 @@
 		power_failure()
 		return FALSE
 
-// Used by child types if they have other power source than battery or recharger
+///Used by subtypes for special cases for power usage
```
Comment (Mothblocks) [inline code/modules/modular_computers/computers/item/computer_power.dm:57]: Document what it returns

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/modular_computers/computers/item/computer.dm (hunk 5 above)
Comment (JohnFulpWillard) [reply code/modules/modular_computers/computers/item/computer.dm:323]: to be honest i didnt touch this, it was just copied from shared computer parts, so i have no clue

File: code/modules/modular_computers/computers/item/computer.dm (hunk 6 above)
Comment (JohnFulpWillard) [reply code/modules/modular_computers/computers/item/computer.dm:453]: image

Tablets aren't improper or anything at the moment, so this is the right way

File: code/modules/modular_computers/computers/item/computer.dm (hunk 3 above)
Comment (JohnFulpWillard) [reply code/modules/modular_computers/computers/item/computer.dm:309]: my plan is to kill all these off when i remove ID holders, but sure

**Comment by JohnFulpWillard** [comment]
i forgot to push

File: code/modules/modular_computers/computers/item/computer.dm (hunk 6 above)
Comment (Mothblocks) [reply code/modules/modular_computers/computers/item/computer.dm:453]: oh yeah

File: code/modules/modular_computers/computers/item/computer.dm (hunk 5 above)
Comment (Mothblocks) [reply code/modules/modular_computers/computers/item/computer.dm:323]: well if you examine and its not plain text then its right

**Pull request merged**